### PR TITLE
Adding Infrastructure and post build step to run the linker on our OOB assemblies

### LINKED ISSUE #48488: Enable ILLink warning analysis for OOB libraries
In #38033 we enabled ILLink warning analysis for assemblies that are in the NETCoreApp shared fx / runtimepack.

With #48428 we are marking all libraries built in `dotnet/runtime` as "trimmable". We should enable ILLink warning analysis for these libraries as well. And then once we are finished resolving the warnings in the libraries in the runtimepack (#45623), we will start working on the rest of the libraries as well.

cc @marek-safar @joperezr @sbomer

### PR DESCRIPTION
**Pull request by joperezr**
cc: @eerhardt @sbomer @krwq @vitek-karas @ViktorHofer 

fixes https://github.com/dotnet/runtime/issues/48488

This is mostly ready for review except that I'm still hitting https://github.com/mono/linker/issues/1416 and trying to find the best way to suppress these warnings which for now are causing the build to fail. That means this PR isn't ready to go in, but sending it out for now to get initial feedback on the infrastructure.

This approach is mostly following the advise we are giving library devs here: https://github.com/dotnet/docs/blob/main/docs/core/deploying/prepare-libraries-for-trimming.md. We are only adding a few additional targets that are accounting for the fact that we have a slightly complex build system, but other than that this is mostly following the guidance on those docs provided by @sbomer

### CODE REVIEW

File: src/libraries/pretest.proj [hunk 1]
```diff
@@ -138,5 +138,20 @@
                              RootAttributes="@(FrameworkListRootAttribute)" />
   </Target>
 
+  <Target Name="ILLinkTrimOOBLibraries"
+          AfterTargets="GenerateRuntimeListFile"
+          DependsOnTargets="SetCommonILLinkArgs"
+          Condition="'$(BuildingNETCoreAppVertical)' == 'true'">
+    <MSBuild Targets="Restore"
+             Projects="$(MSBuildThisFileDirectory)trimming\trimming.csproj"
+             Properties="$(TraversalGlobalProperties);MSBuildRestoreSessionId=$([System.Guid]::NewGuid())" />
```
Comment (ViktorHofer) [inline src/libraries/pretest.proj:147]: Please avoid this extra restore invocation. We worked really hard in the past months / year to reduce extra invocations. If you put the trimming app under a "src" sub-directory it will be restored automatically. Another option would be to add a P2P in pretest.proj trimming.csproj.

File: src/libraries/pretest.proj [hunk 2]
```diff
@@ -138,5 +138,20 @@
                              RootAttributes="@(FrameworkListRootAttribute)" />
   </Target>
 
+  <Target Name="ILLinkTrimOOBLibraries"
```
Comment (ViktorHofer) [inline src/libraries/pretest.proj:141]: What's the reason that you aren't doing this in src.proj where we also trim the inbox ones?

Comment (ViktorHofer) [reply src/libraries/pretest.proj:141]: Is it that you need a ready to use runtime pack?

File: src/libraries/trimming/trimming.csproj [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    <ProjectExclusions Include="$(LibrariesProjectRoot)Microsoft.NETCore.Targets\src\*.csproj" />
+    <!--Exclude single Tests project that doesn't follow default naming conventions of the repo-->
+    <ProjectExclusions Include="$(LibrariesProjectRoot)Microsoft.Extensions.DependencyInjection.Specification.Tests\src\*.csproj" />
+    <!--Exclude Source Generators-->
+    <ProjectExclusions Include="$(LibrariesProjectRoot)Microsoft.XmlSerializer.Generator\src\*.csproj" />
+    <!--Exclude Common Project-->
+    <ProjectExclusions Include="$(LibrariesProjectRoot)Common\src\*.csproj" />
+
+
+    <Compile Include="Program.cs" />
```
Comment (ViktorHofer) [inline src/libraries/trimming/trimming.csproj:35]: Is that file actually needed? Could this be a .proj instead that uses the Traversal SDK similar to src.proj?

File: src/libraries/pretest.proj (hunk 2 above)
Comment (joperezr) [reply src/libraries/pretest.proj:141]: We do the shared framework on src.proj because we are manually invoking the linker and editing the command that we run, while for the OOBs we are instead following [the steps we are asking third party library devs to follow](https://github.com/dotnet/docs/blob/main/docs/core/deploying/prepare-libraries-for-trimming.md) but for that to work, we need to have the RuntimeList.xml file already generated in order to be able to build trimming.csproj against the produced shared framework, and that only happens after pretest.proj.

File: src/libraries/trimming/trimming.csproj (hunk 3 above)
Comment (joperezr) [reply src/libraries/trimming/trimming.csproj:35]: Needs to be a console app in order to be able to be fully trimmed, and if it is a console app, then it needs to have a main, so yes the file is needed.

File: src/libraries/pretest.proj (hunk 1 above)
Comment (joperezr) [reply src/libraries/pretest.proj:147]: I don't think adding a P2P is the right thing to do in this case, specially since it is a Console application project and not a library, but I'll see how I can avoid this manual restore call.

File: src/libraries/pretest.proj (hunk 2 above)
Comment (joperezr) [reply src/libraries/pretest.proj:141]: >Is it that you need a ready to use runtime pack?

Yup, exactly

File: src/libraries/System.Private.DataContractSerialization/src/ILLink/ILLink.Suppressions.xml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<linker>
+  <assembly fullname="System.Private.DataContractSerialization, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
```
Comment (eerhardt) [inline src/libraries/System.Private.DataContractSerialization/src/ILLink/ILLink.Suppressions.xml:3]: I don't understand this. I thought you fixed this assembly in [https://github.com/dotnet/runtime/pull/50619](https://github.com/dotnet/runtime/pull/50619). Why are we getting these now?

File: src/libraries/System.Configuration.ConfigurationManager/src/ILLink/ILLink.Suppressions.xml [hunk 5]
```diff
@@ -0,0 +1,161 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<linker>
+  <assembly fullname="System.Configuration.ConfigurationManager, Version=6.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51">
```
Comment (eerhardt) [inline src/libraries/System.Configuration.ConfigurationManager/src/ILLink/ILLink.Suppressions.xml:3]: ConfigurationManager is not trimmable. See [https://github.com/dotnet/runtime/pull/49085](https://github.com/dotnet/runtime/pull/49085). We probably don't need to waste time analyzing this assembly until we try to make it trimmable (which isn't in the near future).

File: src/libraries/pretest.proj (hunk 1 above)
Comment (ViktorHofer) [reply src/libraries/pretest.proj:147]: That doesn't prevent the P2P. We do have a lot P2Ps to executables i.e. the test runners. Just make sure to set `ReferenceOutputAssembly` to false.

File: src/libraries/pretest.proj (hunk 2 above)
Comment (ViktorHofer) [reply src/libraries/pretest.proj:141]: Hmm does that mean we can't restore before the runtime pack is available? If yes, did we consider treating this as a test instead of the product build? I.e. sourcebuild likely doesn't care about this logic.

File: src/libraries/System.ComponentModel.TypeConverter/src/ILLink/ILLink.Suppressions.xml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<linker>
+  <assembly fullname="System.ComponentModel.TypeConverter, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/ILLink/ILLink.Suppressions.xml:3]: Actually, in a perfect world, we shouldn't be analyzing the `System.ComponentModel.TypeConverter` library at all during the OOB trim analysis run.

**Comment by eerhardt** [comment]
There are a few assemblies here that I think we should set `IsTrimmable=false` on, and not analyze them at all. Here is my initial list I think we should skip trimming for:

* System.CodeDom
* System.ComponentModel.Composition*
* System.Composition*
* System.Configuration.ConfigurationManager (already has IsTrimmable=false on the assembly)
* System.ServiceModel.Syndication ?
* System.Speech

File: src/libraries/System.Private.DataContractSerialization/src/ILLink/ILLink.Suppressions.xml (hunk 4 above)
Comment (joperezr) [reply src/libraries/System.Private.DataContractSerialization/src/ILLink/ILLink.Suppressions.xml:3]: I think this is because the wrong configuration of System.Private.DataContractSerialization is getting picked. One more reason why not to go with this approach I guess.

**Comment by joperezr** [comment]
Closing this as this has been superseded by https://github.com/dotnet/runtime/pull/52272
